core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.AlluxioURI;
+import alluxio.exception.ExceptionMessage;
+import alluxio.exception.InvalidPathException;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.util.io.PathUtils;
+
+import io.netty.util.internal.chmv8.ConcurrentHashMapV8;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class caches paths in the UFS which do not exist.
+ */
+@ThreadSafe
+public class UfsAbsentPathCache {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsAbsentPathCache.class);
+
+  /** The mount table. */
+  protected final MountTable mMountTable;
+  /** Stores a cache for each mount point (the key is mount id). */
+  protected final ConcurrentHashMapV8<Long, ConcurrentSkipListSet<String>> mCaches;
```

Comment:
it seems like this will grow without a bound ... I suggest we only use a single cache, leveraging Guava's cache implementation (https://github.com/google/guava/wiki/CachesExplained) ... entries in the cache are key by UFS URI and store mount point ID as value so that they can be validated / invalidated lazily at lookup (as opposed to invalidating all entries when a mount point is removed)